private-templates-service/server/src/db/providers/StorageProvider.ts

```diff
@@ -0,0 +1,175 @@
+        _id?: string;
+        team: string[] = new Array();
+        org: string[] = new Array();
+        email: string;
+
+        
+        constructor(team: string | string[], org: string | string[], email:string, _id?: string,) {
+            this.team = this.team.concat(team);
+            this.org = this.org.concat(org);
+            this.email = email;
+            if(_id) {
+                this._id = _id;
+            }
+        }
+
+        static fromJSON(j: string) : Collections.User | null{
+            try {
+                let user: Interface.IUser = JSON.parse(j);
+                return new User(user.team, user.org, user.email, user._id);
+            } catch (e) {
+                logger.error("Error with the JSON structure of the User object.")
+                return null;
+            }
+        }
+    }
+
+    export class TemplateInstance implements Interface.ITemplateInstance {
+        json: string;        
+        version: string;
+        constructor(j: string, version: string) {
+            try {
+                this.json = JSON.parse(j);
+            } catch(e) {
+                logger.error("JSON field of the template: " + j);
+                throw(new Error("JSON field of the template instance can't be parsed: invalid structure or empty string:\n" + e))
+            }
+            this.version = version;
+        }
+        static fromJSON(j:string) : Collections.TemplateInstance | null {
+            try {
+                let instance: Interface.ITemplateInstance = JSON.parse(j);
+                return new Collections.TemplateInstance(JSON.stringify(instance.json), instance.version);
+            } catch(e) {
+                logger.error("Error with the JSON structure of the template instance. Object can not be parsed:\n" + e)
+                logger.error("Template Instance JSON: + " + j);
+                return null;
+            }
+        }
+
+        
+    }
+    
+    export class Template implements Interface.ITemplate {
+        _id?: string;
+        instances: Collections.TemplateInstance[] = new Array()
+        tags: string[] = new Array();
+        owner: string;
+        createdAt: Date;
+        updatedAt: Date;
+        isPublished: boolean;   
+        
+        
+        // permission: []
+        constructor(instances: Collections.TemplateInstance[] | Collections.TemplateInstance, tags: string[] | string,
+                                 owner: string, createdAt: Date, 
+                                 updatedAt: Date, isPublished: boolean, _id?: string) {
+            this.instances = this.instances.concat(instances);
+            this.tags = this.tags.concat(tags);
+            this.owner = owner;
+            this.createdAt = createdAt;
+            this.updatedAt = updatedAt;
+            this.isPublished = isPublished;
+            if(_id) {
+                this._id = _id;
+            }
+        }
+        static fromJSON(j: string) : Collections.Template | null {
```

Comment:
For future reference:

A pattern I've seen alot & like - instead of returning the _explicit interface | null_ - you could create an additional interface called 'JSONResponse' (for example). It might look like:

```
export interface JSONResponse {
  success: boolean;
  errorMessage?: string;
  result?: any; // Or you could list the 3 possible classes here.
} 
```

Then this would be the return type. The pros are that the caller could read the 'success' property instead of being forced to check for null, they receive the errorMsg if anything goes wrong, and this interface is easily scalable if their are additional properties you want to include later.